### PR TITLE
Added a basic log browsing UI

### PR DESCRIPTION
**Pull request by superbobry**
The implementation is pretty sketchy right now. Specifically, we can futher

* cover the backend part with tests;
* alter the JS part to query the YARN API directly for the container metadata instead of proxying it via Skein;
* ~use a templating library either on the server-side (e.g. JSP) or on the client-side~;
* ~expose K/V store content~;
* expose scale/kill actions in the UI.

Demo:

screen shot 2018-07-19 at 19 33 39

### CODE REVIEW

**Comment by superbobry** [comment]
I will iterate a bit more on the UI. Not 100% happy with the K/V table right now.

screen shot 2018-07-24 at 14 06 57

**Comment by superbobry** [comment]
Hmm, it looks like the failures are random? /cc @jcrist

File: java/pom.xml [hunk 1]
```diff
@@ -148,6 +149,37 @@
...
+    </dependency>
+
+    <dependency>
+      <groupId>org.webjars</groupId>
+      <artifactId>bootstrap</artifactId>
+      <version>4.1.1</version>
+    </dependency>
+
+    <dependency>
+      <groupId>com.github.spullara.mustache.java</groupId>
```
Comment (jcrist) [inline java/pom.xml]: Not knowing anything about java web development, can you comment on the decision to use mustache and bootstrap here?

Comment (superbobry) [reply java/pom.xml]: To be honest, I don't have much to say either. I needed a minimalistic server-side templating engine with ~readable syntax. I used Mustache.js a while ago on the client-side, so I looked up a Java port. 

There're other alternatives, e.g:
* [Freemarker](https://freemarker.apache.org/),
* [Jtwig](http://jtwig.org/) which seems to be close to Django/Jinja2 wrt syntax.

Personally, I like Jtwig more than Mustache, but I'd go with whatever library you find more appropriate.

Comment (superbobry) [reply java/pom.xml]: I've drafted a Jtwig version and tbh it is not much nicer than the Mustache one. I can push the commit to this branch if you'd like to compare.

**Comment by superbobry** [comment]
screen shot 2018-08-01 at 16 05 17

**Comment by jcrist** [comment]
I just took some time to build and try this out. A few assorted thoughts:

- I'm not sure if/how we should expose the key-value store information. Since the values can be arbitrary bytes, or large in length, displaying them in the UI seems potentially error prone. Do you have a use case for viewing the kv store?

- If the key-value store remains exposed, there may be a large number of values. Perhaps making `services` and `key-value store` separate pages, and making the key-value store a pageable view would help. Alternatively, server-side templating could be dropped and replaced with a simple REST api to pull in the values as needed. No strong thoughts here, I am not a web programmer.

If you have thoughts on the above I'd appreciate them. I hope to hack on this in some free time over the next week.

**Comment by sodre** [comment]
It might be worth looking at the Hashicorp Consul UI for additional inspiration on displaying services and KV pairs.

**Comment by superbobry** [comment]
@jcrist I've exposed the K/V store to ease debugging. My application uses it to implement a barrier as mentioned in #6. The values human-readable, therefore rendering them as-is is not a problem. I agree that in general, this could be problematic. An ad-hoc fix could be to check if the value is printable and render it as-is or base64-encoded.

Regarding client-side rendering, I used that initially, but decided to got for a single-page rendered server-side since it is conceptually simpler. I agree that separate paginated or lazily-loaded views are much better from UX point of view. Also, I've checked out Consul UI following @sodre suggestion, and I think it'd be wonderful if skein had something similar! If we agree on the set of features and the design, I could find someone more well versed in web dev to implement it :)

**Comment by jcrist** [comment]
>  An ad-hoc fix could be to check if the value is printable and render it as-is or base64-encoded.

That or just indicate somehow that it's not human readable. I personally only see value in accessing human readable values in the ui.

I'm surprised `ByteString.toString` doesn't have a checked exception for failure to convert bytes to the specified charset. Perhaps should use[`toStringUtf8` and `isValidUtf8`](https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/ByteString#toStringUtf8--) instead.

> I used that initially, but decided to got for a single-page rendered server-side since it is conceptually simpler.

I'm fine with getting something simple in and iterating on it later. Server side rendering is fine for now, we can always change later if needed. Perhaps should just split the services and key/value tabs into two separate pages for now to reduce some of the message size. I can see myself refreshing the services tab but not caring as much about the key-value store.

> I think it'd be wonderful if skein had something similar! If we agree on the set of features and the design, I could find someone more well versed in web dev to implement it :)

That'd be great. Lets try to get the simple initial version in first and see how it feels, then can discuss improvements in future issues and PRs.

**Comment by superbobry** [comment]
I've added an `isValidUtf8` check, thanks for the tip.

**Pull request closed without merging**